File: NETCore/Cedrus/Controllers/V1/Models/TemplateModel.cs
using Juniper.Cedrus.Entities;

namespace Juniper.Cedrus.Controllers.V1;

public class TemplateModel
{
    public string TypeStamp => "template";
    public int? Id { get; }
    public string? Name { get; }
    public EntityTypeModel EntityType { get; }
    public PropertyTypeModel[] PropertyTypes { get; }

    public TemplateModel(Template? template)
    {
        Id = template?.Id;
        Name = template?.Name;
        EntityType = new EntityTypeModel(template?.EntityType);
        PropertyTypes = template
            ?.PropertyTypes
            ?.Select(t => new PropertyTypeModel(t))
            ?.ToArray()
            ?? Array.Empty<PropertyTypeModel>();
    }
}

File: JS/cedrus/src/models/TemplateModel.ts
import { isArray } from "@juniper-lib/util";
import { EntityTypeModel, isEntityTypeModel } from "./EntityTypeModel";
import { INamed, isINamed } from "./INamed";
import { ITypeStamped, isTypeStamped } from "./ITypeStamped";
import { PropertyTypeModel } from "./PropertyTypeModel";

export interface TemplateModel extends INamed, ITypeStamped<"template"> {
    entityType: EntityTypeModel;
    propertyTypes: PropertyTypeModel[];
}

export function isTemplateModel(obj: unknown): obj is TemplateModel {
    return isTypeStamped("template", obj)
        && isINamed(obj)
        && "entityType" in obj
        && isEntityTypeModel(obj.entityType)
        && "propertyTypes" in obj
        && isArray(obj.propertyTypes);
}